driver orsay: improve imaging reliability and more

pieleric opened the pull request:
This should mostly complete the support for imaging on the Orsay Physics FIB.
It also fixes a few other issues related to reliability found while testing the image acquisition. 
Now the Odemis GUI support playing the image, adjusting the acquisition setting, adjusting the focus.

There is still one bug in the FIB software which is not handle when changing the RoI up in Y. Still hoping that Orsay Physics can directly fix this issue instead of writing a workaround.

For the reviewers: I recommend checking commit per commit, as it might clarify the changes.

Kleijwegt commented on src/odemis/driver/test/orsay_test.py:
```diff
@@ -2026,132 +2080,265 @@ def setUpClass(cls):
 
     def test_get_varying_dwell_time(self):
-        self.skipTest(f"By default this test is skipped. It test the functionality on changing dwell times. User input is required")
+        # self.skipTest(f"By default this test is skipped. It test the functionality on changing dwell times. User input is required")
```
Probably this comment can go away? (or if changing dwelltimes is not that stable yet it should remain uncommented?)

pieleric replied:
Hi Kornee, nice to have you shinning up here!
The change of dwell time is now stable (as far as I could test). So indeed this line can completely go away.

Kleijwegt commented on src/odemis/driver/orsay.py:
```diff
@@ -1547,41 +1591,42 @@ def disconnect(self):
-        if attr_name is None:
-            attr_name = self._attr_name
-        if attr_name != self._attr_name:
+        # Drop events when something else than the "Actual" attribute changed (eg, Target).
```
Spelling mistake? (**e.g.** Target)

Kleijwegt commented on src/odemis/driver/orsay.py:
```diff
@@ -2250,11 +2310,19 @@ def __init__(self, name, role, parent, **kwargs):
+
+        # It uses the pixel convention: smallest value corresponds to top left of the image.
+        # The unit is the pixel at scale == 1.
+        # We accept floats, because it's Odemis convention, however, it will always be rounded
```
I see you use "round" to round the float. I think it is nice to add that info also here so that a reader knows using which function the float is rounded to an int.

tepals commented on src/odemis/driver/orsay.py:
```diff
@@ -1316,8 +1315,8 @@ def _updateTemperature(self, parameter=None, attr_name="Actual"):
             parameter = self._temperaturePar
-        if float(self._temperaturePar.Actual) == float(self._temperaturePar.Target):
-            logging.debug("Target temperature reached.")
+        # if float(self._temperaturePar.Actual) == float(self._temperaturePar.Target):
+        #     logging.debug("Target temperature reached.")
```
why is this commented?

pieleric replied:
Too many logs. It's still handy to know how to check that the target temp is reached, so I left the code.

pieleric commented on src/odemis/driver/orsay.py:
```diff
@@ -2471,51 +2475,50 @@ def _updateImageFormat(self, parameter=None, attr_name="Actual"):
+            self.imageFormat._value = new_value  # to not call the setter
+            self.imageFormat.notify(new_value)
 
     def _clip_and_set_image_area(self, target_resolution, target_translation):
```
I'd rather not, because it's a private function. I prefer doing just "black box" testing, by only using the standard interface. So it's indirectly tested when testing the .translation and .resolution VAs, but not explicitly tested.

pieleric commented on src/odemis/driver/orsay.py:
```diff
@@ -2584,21 +2585,25 @@ def _updateTranslationResolution(self, parameter=None, attr_name="Actual"):
 
-        self.translation._value = new_translation  # to not call the setter
-        self.resolution._value = new_resolution  # to not call the setter
+        if (self.translation._value != new_translation or
+            self.resolution._value != new_resolution
+        ):
```
Yeah... the problem with `if (` is that it's exactly the size of one indentation... so pretty much any formatting looks odd. I've added a line after the opening bracket.

pieleric commented on src/odemis/driver/orsay.py:
```diff
@@ -2891,133 +2915,214 @@ def __init__(self, name, role, parent, **kwargs):
+        img_size = list(map(int, img_size.split(" ")))
+        if len(img_area) != 4 or len(img_size) != 2:
+            logging.error("Cannot crop, img_area decoded to %s, img_size to %s", img_area, img_size)
+        elif not (0 <= img_area[0] < img.shape[1] and
+                  0 <= img_area[1] < img.shape[0] and
```
Yes, because number has Y as first dim and X as second dim. I'll add a comment.

pieleric commented on src/odemis/driver/test/orsay_test.py:
```diff
@@ -2026,132 +2080,265 @@ def setUpClass(cls):
-            if child.name == CONFIG_FIBBEAM["name"]:
-                cls.fibbeam = child
-            elif child.name == CONFIG_SCANNER["name"]:
+            # if child.name == CONFIG_FIBBEAM["name"]:
+            #     cls.fibbeam = child
```
yes, unused... but of course it can now be entirely dropped.

pieleric commented on src/odemis/driver/test/orsay_test.py:
```diff
@@ -2026,132 +2080,265 @@ def setUpClass(cls):
+        """
+        Check that the image is received at the specified scale
+        """
+        max_res = self.scanner.resolution.range[1]
+
+        res = max_res
```
I'm now using it in more places ;-)

pieleric commented on src/odemis/driver/test/orsay_test.py:
```diff
@@ -2026,132 +2080,265 @@ def setUpClass(cls):
+        res = max_res
+        self.scanner.scale.value = (1.0, 1.0)
+        self.scanner.resolution.value = res
+        self.assertEqual(self.scanner.scale.value, (1, 1))
+        self.assertEqual(self.scanner.resolution.value, res)
```
I don't mind. For now that's more PyCharm than a general case, as in the unittest code, it's just called "first" and "second".

pieleric commented:
I've updated the commits with fixes from the review. (Use the "Compare" button to see all the changes since the previous review).
I've also updated the code to work better for some corner cases with the RoI. Also, in the meantime, the new version of the Orsay Physics server has several issues resolves, which allow to run almost all test cases now. 
I've also added a new commit to adjust the beam shift so that it follows the odemis convention.

FCanberk commented:
@tepals @nandishjpatel is it possible that this PR is overlooked? I kinda need this PR merged for my branch :smile:

tepals reviewed (approved):
one small comment
